hasktorch/src/Torch/Data/Pipeline.hs

```diff
@@ -11,145 +14,127 @@
-                           , DatasetMock(..)
-                           , ConcurrentDataset(..)
+module Torch.Data.Pipeline ( Dataset(..)
+                           , MapStyleOptions(..)
+                           , Sample(..)
+                           , makeListT
+                           , mapStyleOpts
+                           , runContT
                            ) where
 
+import           Control.Applicative (Alternative, (<|>))
 import           Control.Concurrent.Async.Lifted
 import qualified Control.Foldl as L
 import           Control.Monad
-import Control.Applicative (Alternative, (<|>))
-import           Data.Maybe (isJust)
+import           Data.Maybe (fromJust, isJust)
 import           Pipes
 import           Pipes.Concurrent
 import qualified Pipes.Prelude as P
 
 import           Control.Monad.Trans.Control (MonadBaseControl(..))
+import           Control.Monad.Cont (runContT, ContT)
+import           Torch.Data.Internal
+import qualified Torch as Torch
+import Lens.Family (view)
+import Pipes.Group
+import Data.Vector (Vector)
+import qualified Data.Map.Strict as M
+import Data.Set
+import Control.Concurrent.STM
+import qualified Data.Vector as V
+import Control.Monad.Base (MonadBase)
+import System.Random
+import qualified Data.IntMap as I
+import Data.IntMap (IntMap)
+import Control.Arrow (Arrow(first))
+
+data MapStyleOptions = MapStyleOptions { bufferSize :: Int
+                                       , numWorkers :: Int
+                                       , shuffle :: Sample
+                                       }
+
+mapStyleOpts numWorkers = MapStyleOptions { bufferSize = numWorkers
+                                          , numWorkers = numWorkers
+                                          , shuffle = Sequential
+                                          }
+
+data Sample = Sequential | Shuffle StdGen
+
+class (Ord k) => Dataset m dataset k sample | dataset -> sample, dataset -> k  where
+  getItem :: dataset -> k -> m sample
+  keys :: dataset -> Set k
+
+---------------------- Workflow --------------------
+-- - make a new map of keys to TVars of samples, possibly shuffled keys, tracking which keys have been sampled
+-- - create a TQueue of keys (using pipes-concurrency wrapper)
+-- - fork off workers which all pull from the TQueue and sample that key using the dataset,
+--   then update the TVar associated with that key
+-- have a worker waiting for each successive key to be updated in the list of (key, TVar)
+
+makeListT :: forall m dataset k sample r .
+  (MonadIO m, MonadBaseControl IO m,  Dataset m dataset k sample) =>
+  MapStyleOptions ->
+  dataset ->
+  ContT r m  (ListT m (sample, Int))
+makeListT MapStyleOptions{..} dataset = do
+  (keyOutput, keyInput, seal) <- liftIO $ spawn' unbounded
+
+  let retreiveSet = liftIO $ keyTVarSet $ keys @m dataset
+  keyTVarSet <- case shuffle of
+    Sequential -> retreiveSet
+    Shuffle g -> fst . fisherYates g <$> retreiveSet
```

Comment:
It could be nice to return the new random generator in the same continuation so that it is available for the next epoch.
can you make the return type
```
ContT r m (ListT m (sample, Int), Sample)
```
?

Reply:
Good idea!